packages/kbn-handlebars/index.ts

```diff
@@ -312,9 +319,27 @@ class ElasticHandlebarsVisitor extends Handlebars.Visitor {
       this.ast = Handlebars.parse(this.template!);
     }
 
-    this.accept(this.ast);
+    let main: Handlebars.HelperDelegate = () => {
+      const prog = this.generateProgramFunction(this.ast!);
+      return prog(this.context, this.runtimeOptions);
+    };
 
-    return this.output.join('');
+    // Run any decorators that might exist on the root
+    main = this.processDecorators(this.ast, main);
+    this.processedRootDecorators = true;
+
+    const result = main(this.context, options);
+    this.output.push(result);
+
+    // This is a workaround to support decorators that returns a custom
+    // function which itself doesn't return a string, in which case joining as
+    // strings will result in an unexpected output.
```

Comment:
__note:__ I did my best, but couldn't wrap my head around this sentence :slightly_smiling_face: Maybe we need to simplify or have an example or something (if possible).

Reply:
How about this re-written explanation instead:

```
    // We normally expect the return value of `main` to be a string. However,
    // if a decorator is used, the return value can be any type. To match the
    // upstream handlebars project behavior, we want the result of rendering
    // the template to be the literal value returned by the decorator.
    //
    // Since the array will always only contain one element in this case, we
    // just return that single value instead of attempting to join all the
    // array elements as strings.
```

Reply:
It's much clearer, thanks!

Reply:
I ended up completely refactoring this section to better show the intent of what is going on. I think it's much easier to understand now.